src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp

```diff
@@ -2477,6 +2483,201 @@ address StubGenerator::generate_unsafe_copy(const char *name,
 }
 
 
+// Static enum for helper
+enum USM_TYPE {USM_SHORT, USM_DWORD, USM_QUADWORD};
+// Helper for generate_unsafe_setmemory
+//
+// Atomically fill an array of memory using 2-, 4-, or 8-byte chunks
+static void do_setmemory_atomic_loop(USM_TYPE type, Register dest,
+                                     Register size, Register wide_value,
+                                     Register tmp, Label& L_exit,
+                                     MacroAssembler *_masm) {
+  Label L_Loop, L_Tail, L_TailLoop;
+
+  int shiftval = 0;
+  int incr = 0;
+
+  switch (type) {
+    case USM_SHORT:
+      shiftval = 1;
+      incr = 16;
+      break;
+    case USM_DWORD:
+      shiftval = 2;
+      incr = 32;
+      break;
+    case USM_QUADWORD:
+      shiftval = 3;
+      incr = 64;
+      break;
+  }
+
+  // At this point, we know the lower bits of size are zero
+  __ shrq(size, shiftval);
+  // size now has number of X-byte chunks (2, 4 or 8)
+  __ cmpq(size, 8);
+  __ jccb(Assembler::below, L_Tail);
+
+  // Number of (8*X)-byte chunks into rScratch1
+  __ movq(tmp, size);
+  __ shrq(tmp, 3);
```

Comment:
`shr` [sets the zero flag][1], so I think you can just move the jump to after the shift and avoid a separate comparison?

```suggestion  
  // Number of (8*X)-byte chunks into rScratch1
  __ movq(tmp, size);
  __ shrq(tmp, 3);
  __ jccb(Assembler::zero, L_Tail);
```

[1]: https://www.felixcloutier.com/x86/sal:sar:shl:shr#flags-affected

Reply:
Good catch.  Done.